backend/typescript/services/implementations/userServicepg.ts

```diff
@@ -0,0 +1,338 @@
+      }
+
+      User.update({
+        firstName: user.firstName,
+        lastName: user.lastName,
+        role: user.role
+       },
+       {
+        where: { id: userId}
+       });
+
+      try {
+        updatedFirebaseUser = await firebaseAdmin
+          .auth()
+          .updateUser(oldUser.authId, { email: user.email });
+      } catch (error) {
+        // rollback Postgres user updates
+        try {
+          await User.update({
+            firstName: oldUser.firstName,
+            lastName: oldUser.lastName,
+            role: oldUser.role
+           },
+           {
+            where: { id: userId}
+           });
+
+        } catch (postgresError) {
+          const errorMessage = [
+            "Failed to rollback Postgres user update after Firebase user update failure. Reason =",
+            postgresError.message,
+            "Postgres user id with possibly inconsistent data =",
+            oldUser.id,
+          ];
+          Logger.error(errorMessage.join(" "));
+        }
+      }
+        
+    } catch (error) {
+      Logger.error(`Failed to update user. Reason = ${error.message}`);
+      throw error;
+    }
+        
+
+    return {
+      id: userId,
+      firstName: user.firstName,
+      lastName: user.lastName,
+      email: user.email,
+      role: user.role,
+    };
+  }
+
+  async deleteUserById(userId: string): Promise<void> {
+    try {
+      const deletedUser: User | null = await User.findByPk(Number(userId));
+
+      if (!deletedUser) {
+        throw new Error(`userid ${userId} not found.`);
+      }
+
+      const numDestroyed: number = await User.destroy({ where: { id: userId }});
+
+      if (numDestroyed <= 0) {
+        throw new Error(`userid ${userId} was not deleted in Postgres.`);
+      }
+
+      try {
+        await firebaseAdmin.auth().deleteUser(deletedUser.authId);
+      } catch (error) {
+        // rollback user deletion in Postgres
+        try {
+          await User.create({
+            firstName: deletedUser.firstName,
+            lastName: deletedUser.lastName,
+            authId: deletedUser.authId,
+            role: deletedUser.role,
+          });
+        } catch (postgresError) {
+          const errorMessage = [
+            "Failed to rollback Postgres user deletion after Firebase user deletion failure. Reason =",
+            postgresError.message,
+            "Firebase uid with non-existent Postgres record =",
+            deletedUser.authId,
+          ];
+          Logger.error(errorMessage.join(" "));
+        }
+
+        throw error;
+      }
+    } catch (error) {
+      Logger.error(`Failed to delete user. Reason = ${error.message}`);
+      throw error;
+    }
+  }
+
+  async deleteUserByEmail(email: string): Promise<void> {
+    try {
+      const firebaseUser: firebaseAdmin.auth.UserRecord = await firebaseAdmin
+        .auth()
+        .getUserByEmail(email);
+      const deletedUser: User | null = await User.findByPk(Number(firebaseUser.userId));
```

Comment:
Should be finding by `authId == firebaseUser.uid`?

```ts
const deletedUser: User | null = await User.findOne({ 
  where: { authId: firebaseUser.uid },
});
```

Reply:
Fixed.